src/NuGet.Core/NuGet.Protocol.Core.v3/HttpSource/HttpSource.cs

```diff
@@ -0,0 +1,583 @@
+                        promptCredentials = await PromptForCredentials(cancellationToken);
+
+                        if (promptCredentials != null)
+                        {
+                            // The user entered credentials, create a new message handler that includes
+                            // these and retry.
+                            await UpdateHttpClient(promptCredentials);
+                            continue;
+                        }
+                    }
+                    finally
+                    {
+                        _httpClientLock.Release();
+                    }
+                }
+
+                if (promptCredentials != null && HttpHandlerResourceV3.CredentialsSuccessfullyUsed != null)
+                {
+                    HttpHandlerResourceV3.CredentialsSuccessfullyUsed(_baseUri, promptCredentials);
+                }
+
+                return response;
+            }
+        }
+
+        private async Task<ICredentials> PromptForCredentials(CancellationToken cancellationToken)
+        {
+            ICredentials promptCredentials = null;
+
+            if (HttpHandlerResourceV3.PromptForCredentials != null)
+            {
+                try
+                {
+                    // Only one prompt may display at a time.
+                    await _credentialPromptLock.WaitAsync();
+
+                    promptCredentials =
+                        await HttpHandlerResourceV3.PromptForCredentials(_baseUri, cancellationToken);
+                }
+                finally
+                {
+                    _credentialPromptLock.Release();
+                }
+            }
+
+            return promptCredentials;
+        }
+
+        private async Task UpdateHttpClient()
+        {
+            // Get package source credentials
+            var credentials = CredentialStore.Instance.GetCredentials(_baseUri);
+
+            if (credentials == null
+                && !String.IsNullOrEmpty(_packageSource.UserName)
+                && !String.IsNullOrEmpty(_packageSource.Password))
+            {
+                credentials = new NetworkCredential(_packageSource.UserName, _packageSource.Password);
+            }
+
+            if (credentials != null)
+            {
+                CredentialStore.Instance.Add(_baseUri, credentials);
+            }
+
+            await UpdateHttpClient(credentials);
+        }
+
+        private async Task UpdateHttpClient(ICredentials credentials)
+        {
+            if (_httpHandler == null)
+            {
+                _httpHandler = await _messageHandlerFactory();
+            }
+
+            _httpHandler.ClientHandler.Credentials = credentials;
+            _httpHandler.ClientHandler.UseDefaultCredentials = (credentials == null);
+
+            var httpClient = new HttpClient(_httpHandler.MessageHandler);
+
+            // Set user agent
+            UserAgent.SetUserAgent(httpClient, UserAgent.UserAgentString);
```

Comment:
this is not really OOP, the method should be UserAgent.SetUserAgent(httpClient), the UserAgentString is already a member and doesn't need to be repeated (yes I know this is not your change)